pkg/gateway/middleware/authinfo.go

```diff
@@ -0,0 +1,70 @@
+package middleware
+
+import (
+	"net/http"
+	"strconv"
+
+	"github.com/skygeario/skygear-server/pkg/core/auth/authinfo"
+	"github.com/skygeario/skygear-server/pkg/core/auth/authtoken"
+	"github.com/skygeario/skygear-server/pkg/core/db"
+	"github.com/skygeario/skygear-server/pkg/core/model"
+)
+
+const (
+	httpHeaderAuthInfoID       = "x-skygear-auth-userid"
+	httpHeaderAuthInfoVerified = "x-skygear-auth-verified"
+	httpHeaderAuthInfoDisabled = "x-skygear-auth-disabled"
+)
+
+// AuthInfoMiddleware injects auth info headers into the request
+// if x-skygear-access-token is present in the request.
+type AuthInfoMiddleware struct {
+	TokenStore    authtoken.Store `dependency:"TokenStore"`
+	AuthInfoStore authinfo.Store  `dependency:"AuthInfoStore"`
+	TxContext     db.TxContext    `dependency:"TxContext"`
+}
+
+// Handle implements InjectableMiddleware.
+func (m *AuthInfoMiddleware) Handle(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		defer next.ServeHTTP(w, r)
```

Comment:
I wonder if we encounter error, should we still call the next handler?

Reply:
Depends on how you handle the error.

If you write something to the response in the middleware, then the next handler won't be able to change status code and headers. Moreover, the next handler would write data on top of the data you have written in the middleware. This would probably lead unexpected behaviour.

See https://golang.org/pkg/net/http/#ResponseWriter.

Reply:
OK. Let me change to not invoke the next handler if error is detected.